### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive for Object.values

### ISSUE THREAD

**Issue by TkDodo** [typescript-eslint/typescript-eslint#10722]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.3&fileType=.tsx&code=MYewdgzgLgBCBGArAXDA3gXxgXnRgsAFBED0JMA7iAE4DWEMAhgwKYAeADi8FCwCZFQkWADdGAGwCuLBrgDySblAB0YqTIAUCRAEpB4aEwgQW1XnwBqE6bJhqbRmNGoBLMAHMA2gF0A3EVJyADMJExgOEAgXKBcRFn1hGD4XaiVxAE8AQWNTcyt1WwVEJVVrTW0dR2c3Lz8AwiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QeoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const obj: {} = {}

// works as expected
const values = Object.values(obj)
const assertedValues = values as string[];

// false positive
const directlyAssertedValues = Object.values(obj) as string[];
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

There should be no error in the provided playground

### Actual Result

There’s an error in line 8:

```
This assertion is unnecessary since it does not change the type of the expression. 8:32 - 8:62
```

### Additional Info

It seems that when we have `Object.values(obj)` that returns an `Array<unknown>`, and we’d want to assert that to a more specific array, the rule thinks the assertion is unnecessary when it’s actually necessary.

Note that this doesn’t happen when we change `obj` to be `Record<string, unknown>`, so maybe this is specific to the type `{}`. I’m aware this type should generally be avoided, but I’m not sure how this makes a difference here.

The error can also be reproduced by inlining the object: `Object.values({})`.

**Comment by TkDodo** [comment]
Note: I’ve just seen that we can also pass a type parameter directly to `Object.values`:

```
const values = Object.values<string>(obj)
```

so I can work around this issue, but I still think there’s something potentially wrong here.

**Comment by kirkwaiblinger** [comment]
This is the classic unfixable false positive 🙃 https://github.com/typescript-eslint/typescript-eslint/issues/9007#issuecomment-2080161404

**Comment by bradzacher** [comment]
sadly yeah this is just a quirk of the typechecker API.
it's probably "technically" possible for us to solve -- but it would require some gnarly logic around interrogating the function signatures and trying to guess if it looks like (a) the function has at least one type parameter, (b) the type parameter is used only in a return location.

given how often this comes up -- maybe it's worth us tackling this gnarly logic...?

**Comment by kirkwaiblinger** [comment]
I'd be in favor of investigation into the feasibility. And if it doesn't seem realistic slapping a Limitations section in the rule describing this case.

**Comment by JoshuaKGoldberg** [comment]
Coming from #11298, is there a more search-friendly title we can use here? "... false positive for generic functions with multiple signatures" maybe?
